core/unit_test/tools/TestScopedProfileRegion.cpp

```diff
@@ -0,0 +1,77 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#include <Kokkos_Profiling_ScopedProfileRegion.hpp>
+
+#include <gtest/gtest.h>
+
+#include <string>
+#include <stack>
+
+namespace {
+
+std::stack<std::string> test_region_stack;
+
+// NOTE: cannot use lambdas because they can only be converted to function
+// pointers if they don't capture anything
+void test_push_region(char const *label) { test_region_stack.push(label); }
+
+void test_pop_region() { test_region_stack.pop(); }
+
+TEST(defaultdevicetype, scoped_profile_region) {
+  Kokkos::Tools::Experimental::set_push_region_callback(test_push_region);
+  Kokkos::Tools::Experimental::set_pop_region_callback(test_pop_region);
+
+  ASSERT_TRUE(test_region_stack.empty());
+
+  // Unnamed guard!  Profile region is popped at the end of the statement.
+  Kokkos::Profiling::ScopedProfileRegion("bug");
```

Comment:
Because of `[[nodiscard]]`, this now throws a warning.

Reply:
Yes, I would suggest removing it.